File: 13. Roman to Integer/13_Roman_to_Integer.py
roman = {
    'I' : 1,
    'V' : 5,
    'X' : 10,
    'L' : 50,
    'C' : 100,
    'D' : 500,
    'M' : 1000
}

class Solution:
    def romanToInt(self, s: str) -> int:
        if len(s) < 1 or len(s) > 15 :
            return 0

        output = 0
        for i in range(len(s)):
            if i + 1 < len(s) and roman[s[i]] < roman[s[i+1]]:
                output -= roman[s[i]]
            else:
                output += roman[s[i]] 

        return output

# class Solution:
#     def romanToInt(self, s: str) -> int:
#         if 1 > len(s) or len(s) > 15 :
#             return 0

#         output = 0
#         for i in range(len(s)):
#             output += roman[s[i]]
#             if (roman[s[i-1]] < roman[s[i]]) and (i != 0):
#                 output -= 2*roman[s[i-1]]

#         return output

File: Leetcode_problems/13. Roman to Integer/13_Roman_to_Integer.c
int romanToInt(char * s){
    int t['X' + 1] = {
        ['I'] = 1,
        ['V'] = 5,
        ['X'] = 10,
        ['L'] = 50,
        ['C'] = 100,
        ['D'] = 500,
        ['M'] = 1000,
    };
    /* A simple hash table implemented with array
        C : t[67] = 100
        D : t[68] = 500
        I : t[73] = 1
        L : t[76] = 50
        M : t[77] = 1000
        V : t[86] = 5
        X : t[88] = 10
    */

    int output = 0;
    for (int i = 0; s[i]; i++) {
        if (t[s[i]] < t[s[i+1]]){
            output -= t[s[i]];
        }
        else{
            output += t[s[i]];
        }
    }
    return output;
}